src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -69,6 +69,97 @@ def __init__(
+
+        # If we return this ireq, it should trigger uninstallation of the
+        # existing distribution for reinstallation.
+        ireq.should_reinstall = True
+
+        # Reinstall if --force-reinstall is set.
+        if self.factory.force_reinstall:
+            return ireq
+
+        # The artifact represented by the incoming candidate.
+        cand_link = candidate.source_link
+
+        # The candidate does not point to an artifact. This means the resolver
+        # has already decided the installed distribution is good enough.
+        if cand_link is None:
+            return None
+
+        # The incoming candidate was produced only from version requirements.
+        # Reinstall if the installed distribution's version does not match.
+        if not direct_url_requested:
+            if installed_dist.version == candidate.version:
+                return None
+            return ireq
+
+        # At this point, the incoming candidate was produced from a direct URL.
+        # Determine whether to upgrade based on flags and whether the installed
+        # distribution was done via a direct URL.
+
+        # Always reinstall a direct candidate if it's on the local file system.
+        if cand_link.is_file:
+            return ireq
+
+        # Reinstall if --upgrade is specified.
+        if self.upgrade_strategy != "to-satisfy-only":
+            return ireq
+
+        # The PEP 610 direct URL representation of the installed distribution.
+        dist_direct_url = installed_dist.direct_url
+
+        # The incoming candidate was produced from a direct URL, but the
+        # currently installed distribution was not. Always reinstall to be sure.
+        if dist_direct_url is None:
+            return ireq
+
+        # Editable candidate always triggers reinstallation.
+        if candidate.is_editable:
+            return ireq
+
+        # The currently installed distribution is editable, but the incoming
+        # candidate is not. Uninstall the editable one to match.
+        if installed_dist.editable:
+            return ireq
+
+        # Now we know both the installed distribution and incoming candidate
+        # are based on direct URLs, and neither are editable. Don't reinstall
+        # if the direct URLs match.
+        if link_matches_direct_url(cand_link, dist_direct_url):
+            return None
+
+        # One exception for VCS. Try to resolve the requested reference to see
+        # if it matches the commit ID recorded in the installed distribution.
+        # Reinstall only if the resolved commit IDs don't match.
+        cand_vcs = vcs.get_backend_for_scheme(cand_link.scheme)
+        dist_direct_info = dist_direct_url.info
+        if cand_vcs and ireq.source_dir and isinstance(dist_direct_info, VcsInfo):
+            candidate_rev = cand_vcs.get_revision(ireq.source_dir)
+            if candidate_rev != dist_direct_info.commit_id:
+                return ireq
+        return None
```

Comment:
The issue with cached wheel is because there is no `ireq.source_dir` when the link was found in the wheel cache.

We can resolve this and simplify by doing this (assuming a `DirectUrl.__eq__` that considers `commit_id` and ignores `requested_revision`):

```suggestion
        cand_direct_url = direct_url_from_link(
            cand_link, ireq.source_dir, ireq.original_link_is_in_wheel_cache
        )
        if cand_direct_url == dist_direct_url:
            return None

        return ireq
```

Because `direct_url_from_link` does already have the necessary logic to look up `commit_id` in `source_dir` if needed.

Reply:
@uranusjr I'm proposing this in https://github.com/uranusjr/pip/pull/1